File: src/Winium.Elements.Desktop/ComboBox.cs
namespace Winium.Elements.Desktop
{
    #region using

    using System.Collections.Generic;
    using System.Linq;
    using System.Text.RegularExpressions;

    using OpenQA.Selenium;
    using OpenQA.Selenium.Remote;

    #endregion

    public class ComboBox : DesktopElement
    {
        #region Constants

        private const string CollapseComboBox = "collapseComboBox";

        private const string ExpandComboBox = "expandComboBox";

        private const string FindComboBoxSelectedItem = "findComboBoxSelectedItem";

        private const string IsComboBoxExpanded = "isComboBoxExpanded";

        private const string ScrollToComboBoxItem = "scrollToComboBoxItem";

        #endregion

        #region Constructors and Destructors

        static ComboBox()
        {
            CommandInfoRepository.Instance.TryAddCommand(
                IsComboBoxExpanded,
                new CommandInfo("POST", "/session/{sessionId}/element/{id}/combobox/expanded"));

            CommandInfoRepository.Instance.TryAddCommand(
                ExpandComboBox,
                new CommandInfo("POST", "/session/{sessionId}/element/{id}/combobox/expand"));

            CommandInfoRepository.Instance.TryAddCommand(
                CollapseComboBox,
                new CommandInfo("POST", "/session/{sessionId}/element/{id}/combobox/collapse"));

            CommandInfoRepository.Instance.TryAddCommand(
                FindComboBoxSelectedItem,
                new CommandInfo("POST", "/session/{sessionId}/element/{id}/combobox/selected/element"));

            CommandInfoRepository.Instance.TryAddCommand(
                ScrollToComboBoxItem,
                new CommandInfo("POST", "/session/{sessionId}/element/{id}/combobox/scroll"));
        }

        public ComboBox(IWebElement element)
            : base(element)
        {
        }

        #endregion

        #region Public Properties

        public bool IsExpanded
        {
            get
            {
                var parameters = new Dictionary<string, object> { { "id", this.Id } };
                var response = this.Execute(IsComboBoxExpanded, parameters);

                return bool.Parse(response.Value.ToString());
            }
        }

        #endregion

        #region Public Methods and Operators

        public void Collapse()
        {
            var parameters = new Dictionary<string, object> { { "id", this.Id } };
            this.Execute(CollapseComboBox, parameters);
        }

        public void Expand()
        {
            var parameters = new Dictionary<string, object> { { "id", this.Id } };
            this.Execute(ExpandComboBox, parameters);
        }

        public RemoteWebElement GetSelected(int row, int column)
        {
            var parameters = new Dictionary<string, object> { { "id", this.Id } };
            var response = this.Execute(FindComboBoxSelectedItem, parameters);

            return this.CreateRemoteWebElementFromResponse(response);
        }

        public RemoteWebElement ScrollTo(By by)
        {
            var match = Regex.Match(by.ToString(), @"\.(.*)\: (.*)");
            var strategy = Regex.Replace(match.Groups[1].Value, "([A-Z])", " $1").Split('[').First().Trim().ToLower();
            var value = match.Groups[2].Value;

            var response = this.Execute(
                ScrollToComboBoxItem,
                new Dictionary<string, object> { { "id", this.Id }, { "using", strategy }, { "value", value }, });

            return this.CreateRemoteWebElementFromResponse(response);
        }

        #endregion
    }
}

File: java/src/main/java/winium/elements/desktop/ComboBox.java
package winium.elements.desktop;

import org.openqa.selenium.By;
import org.openqa.selenium.WebElement;
import org.openqa.selenium.remote.RemoteWebElement;
import org.openqa.selenium.remote.Response;

import winium.elements.desktop.extensions.ByExtensions;

import java.util.HashMap;

public class ComboBox extends DesktopElement {
    static final String COLLAPSE_COMBO_BOX = "collapseComboBox";
    static final String EXPAND_COMBO_BOX = "expandComboBox";
    static final String FIND_COMBO_BOX_SELECTED_ITEM = "findComboBoxSelectedItem";
    static final String IS_COMBO_BOX_EXPANDED = "isComboBoxExpanded";
    static final String SCROLL_TO_COMBO_BOX_ITEM = "scrollToComboBoxItem";

    private Response callComboBoxCommand(String command)
    {
        HashMap<String, Object> parameters = new HashMap<String, Object>();
        parameters.put("id", this.getId());
        return this.execute(command, parameters);
    }

    public ComboBox(WebElement element) {
        super(element);
    }

    public boolean isExpanded() {
        HashMap<String, Object> parameters = new HashMap<String, Object>();
        parameters.put("id", this.getId());
        Response response = this.execute(IS_COMBO_BOX_EXPANDED, parameters);

        return Boolean.parseBoolean(response.getValue().toString());
    }

    public void collapse() {
        this.callComboBoxCommand(COLLAPSE_COMBO_BOX);
    }

    public void expand() {
        this.callComboBoxCommand(EXPAND_COMBO_BOX);
    }

    public RemoteWebElement findSelected() {
        return this.createRemoteWebElementFromResponse(this.callComboBoxCommand(FIND_COMBO_BOX_SELECTED_ITEM));
    }

    public RemoteWebElement scrollTo(By by) {
        HashMap<String, Object> parameters = new HashMap<String, Object>();
        parameters.put("id", this.getId());
        parameters.put("using", ByExtensions.getStrategy(by));
        parameters.put("value", ByExtensions.getValue(by));
        Response response = this.execute(SCROLL_TO_COMBO_BOX_ITEM, parameters);
        return this.createRemoteWebElementFromResponse(response);
    }
}
